README.md
# zap-me

With this library, you can operate various electric shock collars remotely using
your Arduino.

**WARNING**: E-Stim is not without risk. Make sure you know about the safety
rules and use this library responsibly and consensually. I am not responsible
for any damage caused by this library or its users.

You are **not** allowed to use this library with animals.

## Currently supported devices

This library currently supports the following devices:

* Chinese Shock Collar often designated with model number 880-3 Sometimes,
it is also labeled as "1000M IP6X". It has a characteristic remote with an
antenna almost as long as the remote, an LED next to it, supports 3 channels
and 3 functions (shock, vibration, sound).

* Dogtronic MT500 (experimental). This is a very old and expensive
collar that was made in Germany by Dogtronic (formerly known as Petsystems).
It features a 3-button remote with a small display and comes in an audio and
a vibration edition.

## Required Hardware

For the CH8803 and the Dogtronic collars you need a 433 Mhz transmitter, such as
the MX-FS-03V. These transmitters usually have 3 pins (1 for data, 1 GND, 1 VCC)
and connecting it to an Arduino is really easy: Just connect the data pin to
your designated transmit pin (any digital pin, e.g. PIN 10) and VCC/GND
according to specifications.

**Important**: If you want to achieve more range, you will have to solder an antenna
to the transmitter. Without an antenna, the range is still several meters in-door
in my tests though.

## FAQ

Q: Can I sniff my remote id with this library?

A: Not yet. Sniffing the id is possible with a MX-05V receiver but you will have
   to decode the data yourself. In the future, we might add a decoder to allow
   simultaneous use of the real remote and this library.

Q: Why is there no boost functionality for the Dogtronic collars?

A: There is no such function built into the collar. Hitting the boost button
   simply sends a continuous signal that is 2 levels higher than the current
   remote setting.

Q: Why can't I chose a random id with Dogtronic?

A: We don't know (yet). With the current implementation, we have found that
   only four ids work, see the comment in the header for more information.


CHANGELOG.md
# Change Log

## [1.1] - 2022-02-06

### Added

- Support for PaiPaitek contributed by LTX.

## [1.0] - 2020-11-14

Initial release with support for 880-3 / 1000M IP6X and MT500.


ZapMe.cpp
/*
  ZapMe - Arduino libary for remote controlled electric shock collars.

  BSD 2-Clause License

  Copyright (c) 2020, sd_ <sd@smjg.org> and others.
  Please see the AUTHORS file in the main source directory for details.
  All rights reserved.

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions are met:

  1. Redistributions of source code must retain the above copyright notice, this
     list of conditions and the following disclaimer.

  2. Redistributions in binary form must reproduce the above copyright notice,
     this list of conditions and the following disclaimer in the documentation
     and/or other materials provided with the distribution.

  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

#include "ZapMe.h"

void ZapMe::sendTiming(uint16_t* timings) {
  if (!timings)
    return;

  /* We start at low */
  bool tlev = false;

  /* Loop until we hit NULL in the timings data */
  for (uint32_t tidx = 0; timings[tidx]; ++tidx, tlev=!tlev) {
    digitalWrite(mTransmitPin, tlev ? HIGH : LOW);
    delayMicroseconds(timings[tidx]);
  }

  /* Always pull down before leaving */
  digitalWrite(mTransmitPin, LOW);
}

void CH8803::sendShock(uint8_t strength, uint16_t duration) {
  DEBUG_ZAPME_MSGLN("CH8803::sendShock");
  sendCommand(1, strength, duration);
}

void CH8803::sendVibration(uint8_t strength, uint16_t duration) {
  DEBUG_ZAPME_MSGLN("CH8803::sendVibration");
  sendCommand(2, strength, duration);
}

void CH8803::sendAudio(uint8_t strength /* unused */, uint16_t duration) {
  DEBUG_ZAPME_MSGLN("CH8803::sendAudio");
  sendCommand(3, 0, duration);
}

void CH8803::sendCommand(uint8_t func, uint8_t strength, uint16_t duration) {
  /* Calculate checksum */
  uint8_t checksum = 0;
  checksum += (uint8_t)(mId >> 8);
  checksum += (uint8_t)(mId & 0xFF);
  checksum += mChannel;
  checksum += func;
  checksum += strength;

  const uint16_t pulseLength = 1016;
  const uint16_t zeroLength = 292;
  const uint16_t oneLength = 804;

  uint16_t transmitIdx = 0;

  /* This is the sync preamble */
  mTransmitTimings[transmitIdx++] = 840;
  mTransmitTimings[transmitIdx++] = 1440;
  mTransmitTimings[transmitIdx++] = pulseLength - zeroLength;

  TRBITS(mId, 16);
  TRBITS(mChannel, 4);
  TRBITS(func, 4);
  TRBITS(strength, 8);
  TRBITS(checksum, 8);

  // Trail, which is 3 zeros, but the third zero has a longer pulse
  TRBITS((uint8_t)0, 2);
  mTransmitTimings[transmitIdx++] = zeroLength;
  mTransmitTimings[transmitIdx++] = 1476;

  if (mMaxTransmitTimings <= transmitIdx) {
    // This should never happen
    DEBUG_ZAPME_MSG("ERROR: Exceeding maximum allocated transmit timings: ");
    DEBUG_ZAPME_MSG(transmitIdx);
    DEBUG_ZAPME_MSGLN(mMaxTransmitTimings);
    return;
  }

  /* Null terminate */
  mTransmitTimings[transmitIdx++] = 0;

  uint32_t startTime = millis();

  #ifdef DEBUG_ZAPME
  DEBUG_ZAPME_MSG("The following timings will be transmitted: ");
  for (uint32_t tidx = 0; mTransmitTimings[tidx]; ++tidx) {
    DEBUG_ZAPME_MSG(mTransmitTimings[tidx]);
    DEBUG_ZAPME_MSG(",");
  }
  DEBUG_ZAPME_MSGLN("0");
  #endif

  DEBUG_ZAPME_MSG("Starting transmission...");

  do {
    /* Transmit timings */
    sendTiming(mTransmitTimings);
    DEBUG_ZAPME_MSG(".");
  } while(millis() - startTime < duration);

  DEBUG_ZAPME_MSGLN(" complete.");
}

void DogTronic::sendShock(uint8_t strength, uint16_t duration) {
  DEBUG_ZAPME_MSGLN("DogTronic::sendShock");
  sendCommand(0, strength, duration);
}

void DogTronic::sendVibration(uint8_t strength /* unused */, uint16_t duration) {
  DEBUG_ZAPME_MSGLN("DogTronic::sendVibration");
  sendCommand(0, 0, duration);
}

void DogTronic::sendAudio(uint8_t strength /* unused */, uint16_t duration) {
  DEBUG_ZAPME_MSGLN("DogTronic::sendAudio");
  sendCommand(0, 0, duration);
}

void DogTronic::sendCommand(uint8_t func /* unused */, uint8_t strength, uint16_t duration) {
  const uint16_t pulseLength = 2212;
  const uint16_t oneGap = 8144;
  const uint16_t zeroGap = 4012;
  const uint16_t endGap = 64000;

  const uint16_t checksum_base = 4; // 0b0100
  const uint8_t unknown_const = 2;  // 0b10

  uint16_t transmitIdx = 0;

  /* This is the sync preamble */
  mTransmitTimings[transmitIdx++] = 240;
  mTransmitTimings[transmitIdx++] = 1700;
  for (uint8_t pi = 0; pi < 14; ++pi) {
    mTransmitTimings[transmitIdx++] = 240;
    mTransmitTimings[transmitIdx++] = 776;
  }
  mTransmitTimings[transmitIdx++] = 388;
  mTransmitTimings[transmitIdx++] = pulseLength;

  /* Translate l bits to timings */
  #define TRBITS_DT(b,l)                              \
    for (uint8_t k = l - 1; k <= l - 1; --k) {        \
      uint16_t bitGap = zeroGap;                      \
      if (((b) & ( 1 << k )) >> k)                    \
        bitGap = oneGap;                              \
      mTransmitTimings[transmitIdx++] = bitGap;       \
      mTransmitTimings[transmitIdx++] = pulseLength;  \
    };

  uint16_t command = 0;

  /*
     b15 b14 b13 b12 b11 b10 b9 b8 b7 b6 b5 b4 b3 b2 b1 b0
     |-------- id ---------| |-- str --| |-2-| |-- chk --|
  */

  /* First 6 bits are the ID */
  command |= (mId << 10);

  /* We now need to convert the strength from MSB to LSB */
  if (strength & 1)
    command |= 1 << 9;

  if (strength & (1 << 1))
    command |= 1 << 8;

  if (strength & (1 << 2))
    command |= 1 << 7;

  if (strength & (1 << 3))
    command |= 1 << 6;

  /* Now we have the unknown 2-bit constant */
  command |= unknown_const << 4;

  /* Calculate the checksum from the base and strength */
  uint16_t checksum = (checksum_base + strength) % 16;

  /* If an overflow happened, feed the overflow back into the sum */
  checksum += (checksum_base + strength) >> 4;

  /* Now write the checksum with bits flipped pairwise */
  if (checksum & 1)
    command |= 1 << 1;

  if (checksum & (1 << 1))
    command |= 1 << 0;

  if (checksum & (1 << 2))
    command |= 1 << 3;

  if (checksum & (1 << 3))
    command |= 1 << 2;


  DEBUG_ZAPME_MSG("The following command will be encoded: ");
  DEBUG_ZAPME_MSGLN(command);

  // Translate bits to timings
  TRBITS_DT(command, 16);

  mTransmitTimings[transmitIdx++] = endGap;

  /* Null terminate */
  mTransmitTimings[transmitIdx++] = 0;

#ifdef DEBUG_ZAPME
  DEBUG_ZAPME_MSG("The following timings will be transmitted: ");
  for (uint32_t tidx = 0; mTransmitTimings[tidx]; ++tidx) {
    DEBUG_ZAPME_MSG(mTransmitTimings[tidx]);
    DEBUG_ZAPME_MSG(",");
  }
  DEBUG_ZAPME_MSGLN("0");
#endif

  DEBUG_ZAPME_MSG("Starting transmission...");

  uint32_t startTime = millis();

  do {
    /* Transmit timings */
    sendTiming(mTransmitTimings);
    DEBUG_ZAPME_MSG(".");
  } while(millis() - startTime < duration);

  DEBUG_ZAPME_MSGLN(" complete.");
}


void PaiPaitek::sendShock(uint8_t strength, uint16_t duration) {
  DEBUG_ZAPME_MSGLN("PaiPaitek::sendShock");
  sendCommand(1, strength, duration);
}

void PaiPaitek::sendVibration(uint8_t strength, uint16_t duration) {
  DEBUG_ZAPME_MSGLN("PaiPaitek::sendVibration");
  sendCommand(2, strength, duration);
}

void PaiPaitek::sendAudio(uint8_t strength /* unused */, uint16_t duration) {
  DEBUG_ZAPME_MSGLN("PaiPaitek::sendAudio");
  sendCommand(4, 0, duration);
}

void PaiPaitek::sendCommand(uint8_t func, uint8_t strength, uint16_t duration) {
  const uint16_t pulseLength = 1000;
  const uint16_t zeroLength = 250;
  const uint16_t oneLength = 750;

  uint16_t transmitIdx = 0;
  uint8_t chSum = 0;

  /* This is the sync preamble */
  mTransmitTimings[transmitIdx++] = 4000;
  mTransmitTimings[transmitIdx++] = 1440;
  mTransmitTimings[transmitIdx++] = 980;

  // Channel translation
  switch (mChannel) {
    case 1: TRBITS(8, 4); chSum = 14; break; // Ch1
    case 2: TRBITS(15, 4); chSum = 0; break; // Ch2
    case 3: TRBITS(10, 4); chSum = 5; break; // Ch3
  }

  TRBITS(func, 4);
  TRBITS(mId, 16);
  TRBITS(strength, 8);

  // Function Checksum
  switch (func) {
    case 1: TRBITS(7, 4); break;  // Shock
    case 2: TRBITS(11, 4); break; // Vibration
    case 4: TRBITS(13, 4); break; // Sound
  }

  // Channel checksum
  TRBITS(chSum, 4);

  // Trail
  mTransmitTimings[transmitIdx++] = zeroLength;
  mTransmitTimings[transmitIdx++] = 1476;

  if (mMaxTransmitTimings <= transmitIdx) {
    // This should never happen
    DEBUG_ZAPME_MSG("ERROR: Exceeding maximum allocated transmit timings: ");
    DEBUG_ZAPME_MSG(transmitIdx);
    DEBUG_ZAPME_MSGLN(mMaxTransmitTimings);
    return;
  }

  /* Null terminate */
  mTransmitTimings[transmitIdx++] = 0;

  uint32_t startTime = millis();

#ifdef DEBUG_ZAPME
  DEBUG_ZAPME_MSG("The following timings will be transmitted: ");
  for (uint32_t tidx = 0; mTransmitTimings[tidx]; ++tidx) {
    DEBUG_ZAPME_MSG(mTransmitTimings[tidx]);
    DEBUG_ZAPME_MSG(",");
  }
  DEBUG_ZAPME_MSGLN("0");
#endif

  DEBUG_ZAPME_MSG("Starting transmission...");

  do {
    /* Transmit timings */
    sendTiming(mTransmitTimings);
    DEBUG_ZAPME_MSG(".");
  } while(millis() - startTime < duration);

  DEBUG_ZAPME_MSGLN(" complete.");
}



ZapMe.h
/*
  ZapMe - Arduino libary for remote controlled electric shock collars.

  BSD 2-Clause License

  Copyright (c) 2020, sd_ <sd@smjg.org> and others.
  Please see the AUTHORS file in the main source directory for details.
  All rights reserved.

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions are met:

  1. Redistributions of source code must retain the above copyright notice, this
     list of conditions and the following disclaimer.

  2. Redistributions in binary form must reproduce the above copyright notice,
     this list of conditions and the following disclaimer in the documentation
     and/or other materials provided with the distribution.

  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

#ifndef _ZapMe_h
#define _ZapMe_h

#if defined(ARDUINO) && ARDUINO >= 100
    #include "Arduino.h"
#endif

#include <stdint.h>

/* To enable debugging, uncomment the next line.
 * You are responsible for enabling Serial communication yourself, e.g.
 * put `Serial.begin(9600);` into your setup routine.
 */
//#define DEBUG_ZAPME 1

#ifdef DEBUG_ZAPME
#define DEBUG_ZAPME_MSGLN(msg) \
  do {                         \
    Serial.println(msg);       \
  } while(false);
#define DEBUG_ZAPME_MSG(msg) \
  do {                       \
    Serial.print(msg);       \
  } while(false);
#else
#define DEBUG_ZAPME_MSGLN(msg) do {} while(false);
#define DEBUG_ZAPME_MSG(msg) do {} while(false);
#endif

/*
 * Common macro to translate bits into timings as follows:
 *
 * Each bit is a pulse of fixed-width (`pulseLength` microseconds), where
 * a zero has a raise of `zeroLength` microseconds at the beginning and
 * a one has a raise of `oneLength` microseconds at the beginning.
 */

 /* Translate l bits to timings */
 #define TRBITS(b,l) \
   for (uint8_t k = l - 1; k <= l - 1; --k) {                   \
     uint16_t bitLength = zeroLength;                           \
     if (((b) & ( 1 << k )) >> k)                               \
       bitLength = oneLength;                                   \
     mTransmitTimings[transmitIdx++] = bitLength;               \
     mTransmitTimings[transmitIdx++] = pulseLength - bitLength; \
   };

class ZapMe {
  public:
    ZapMe(uint8_t transmitPin) : mTransmitPin(transmitPin), mDebug(false) {
      pinMode(mTransmitPin, OUTPUT);
    }

    virtual void sendShock(uint8_t strength, uint16_t duration) = 0;
    virtual void sendVibration(uint8_t strength, uint16_t duration) = 0;
    virtual void sendAudio(uint8_t strength, uint16_t duration) = 0;

    void sendTiming(uint16_t* timings);

  protected:
    uint8_t mTransmitPin;
    bool mDebug;
};


class CH8803 : public ZapMe {
  /*
   * Chinese shockcollar 880-3 (Euro) uses 5 bytes of data per command:
   *
   * | ID (16-bit) | Channel (4 bit) | Function (4 bit) | Strength (8 bit) | Checksum (8 bit) |
   *
   * The ID is simply an identifier randomly assigned to each remote. Collars
   * can be set to learning mode to be associated with a new ID.
   *
   * The channel is an integer that is either 0, 1 or 2 mapping to CH1, CH2 or CH3 on the remote.
   *
   * The function is an integer that is either 1 (Shock), 2 (Vibration) or 3 (Sound).
   *
   * The strength is an integer in range 0 to 99. It is always 0 for sound.
   *
   * The checksum is simply an unsigned sum of all the previous bytes.
   */


  public:
    CH8803(uint8_t transmitPin, uint16_t id)
      : ZapMe(transmitPin), mId(id), mChannel(0) {
        /*
         * In transmit timings, we need 3 timings for the sync preamble,
         * then 64 timings to send 40 bit of data, then 3 trailing zeros
         * and one for the terminating null byte we are going to use as
         * a length indicator.
         */
        mMaxTransmitTimings = 3 + 2*40 + 2*3 + 1;
        mTransmitTimings = new uint16_t[mMaxTransmitTimings];
      }

    void setId(uint16_t id) { mId = id; }
    void setId(uint8_t hid, uint8_t lid) { mId = (hid << 8) + lid; }

    void setChannel(uint8_t ch) { mChannel = ch; }

    virtual void sendShock(uint8_t strength, uint16_t duration);
    virtual void sendVibration(uint8_t strength, uint16_t duration);
    virtual void sendAudio(uint8_t strength, uint16_t duration);

  protected:

    void sendCommand(uint8_t func, uint8_t strength, uint16_t duration);

    uint16_t mId;
    uint8_t mChannel;
    uint16_t* mTransmitTimings;
    uint16_t mMaxTransmitTimings;
};

class DogTronic : public ZapMe {
  /*
   * DogTronic uses a sync preamble of short pulses followed by a relatively
   * slow series of constant length pulses. The gaps inbetween the pulses
   * encode bits (short gap is 0, long gap is 1) and each command has 16 bits.
   *
   * For a command c[15:0], we have:
   *
   * c[15:10] - 6-bit ID (presumably)
   * c[9:6]   - 4-bit shock strength (LSB first)
   * c[5:4]   - 2-bit constant 0b10, unknown purpose
   * c[3:0]   - 4-bit checksum (see below)
   *
   * The checksum c[3:0] is particularly weird of construction:
   *
   * This is a 4-bit value, but the bits are reordered, the proper number
   * is reconstructed as c[2] c[3] c[0] c[1], so this is MSB but the bits
   * have been pairwise swapped. Apart from that, it is a simple sum that
   * does not cover the ID, but only the strength. Since the constant at
   * c[5:4] never changes, we can't tell if it is part of the sum or not.
   * The sum starts at an offset of 0b0100 and then the strength is added.
   * However, if an overflow beyond 4-bit occurs, the sum is not simply
   * truncated (mod 16), but the overflow bit is fed back into the register
   * from the LSB side, so the checksum is:
   *
   *   (0b0100 + strength) mod 16 + (0b0100 + strength) >> 4.
   *
   * Internally, this is probably a 4-bit full adder with the carry-out
   * hard-wired back to the carry-in.
   *
   * IMPORTANT: So far, with this implementation, the collar only accepts
   *            4 different IDs as valid, these are: 14, 23, 44 and 53.
   *            We don't know, why this is the case, it could be tied to
   *            the checksum somehow but without more samples, it is hard
   *            to tell.
   */

  public:
    DogTronic(uint8_t transmitPin, uint8_t id)
      : ZapMe(transmitPin), mId(id) {
        /*
         * In transmit timings, we need 32 timings for the sync preamble,
         * then 32 timings to send 16 bit of data, finally 1 timing for
         * the larger gap after the message and 1 timing for the null byte.
         */
        mMaxTransmitTimings = 32*2 + 2;
        mTransmitTimings = new uint16_t[mMaxTransmitTimings];
      }

    void setId(uint8_t id) { mId = id; }

    virtual void sendShock(uint8_t strength, uint16_t duration);

    /*
     * Dogtronic has no separate vibration and audio, the function is
     * the same but there are two different versions of the collar,
     * one with vibration and one with audio. So it doesn't really matter
     * which of the two here is used, they map to the same functionality.
     */
    virtual void sendVibration(uint8_t strength, uint16_t duration);
    virtual void sendAudio(uint8_t strength, uint16_t duration);

  protected:

    void sendCommand(uint8_t func, uint8_t strength, uint16_t duration);

    uint8_t mId;
    uint16_t* mTransmitTimings;
    uint16_t mMaxTransmitTimings;
};

#endif

class PaiPaitek : public ZapMe {
  /*
   * PaiPaiTek shockcollar (Euro) uses 5 bytes of data per command:
   *
   *  | Channel (4 bit) | Function (4 bit) | ID (16-bit) | Strength (8 bit) | Checksum (8 bit) |
   *
   * The ID is simply an identifier randomly assigned to each remote.
   * Collars can be set to learning mode to be associated with a new ID.
   *
   * The channel is either 8, 15 or 10, mapping to CH1, CH2 or CH3 on the remote.
   * For the `setChannel` API, we map these to 1, 2 and 3 for simplicity.
   *
   * The function is an integer that is either
   *   1 (Shock)
   *   2 (Vibration) or
   *   4 (Sound).
   *
   * The strength is an integer in range 1 to 100. It is always 0 for sound.
   *
   * The checksum consists of 4 bits for the function and 4 bits for channel.
   * Function (upper 4 bit) sum is 11 for Vib, 13 for Sound or 7 for Shock.
   * Channel (lower 4 bit) sum is 14 for CH1, 0 for CH2 and 5 for CH3.
   *
   * The Preambel is 4000 long low / 1480 High / 960 Low.
   *
   * The Ending is 240 High / 2000 Low.
   *
   * The Timing is 250 -> 0 / 750 -> 1 / Complete 1000.
   */

  public:
    PaiPaitek(uint8_t transmitPin, uint16_t id)
      : ZapMe(transmitPin), mId(id), mChannel(0) {
        /*
         * In transmit timings, we need 3 timings for the sync preamble,
         * then 64 timings to send 40 bit of data, then 3 trailing zeros
         * and one for the terminating null byte we are going to use as
         * a length indicator.
         */
        mMaxTransmitTimings = 3 + 2*40 + 3 + 1;
        mTransmitTimings = new uint16_t[mMaxTransmitTimings];
      }

    void setId(uint16_t id) { mId = id; }
    void setId(uint8_t hid, uint8_t lid) { mId = (hid << 8) + lid; }

    void setChannel(uint8_t ch) { mChannel = ch; }

    virtual void sendShock(uint8_t strength, uint16_t duration);
    virtual void sendVibration(uint8_t strength, uint16_t duration);
    virtual void sendAudio(uint8_t strength, uint16_t duration);
    virtual void sendtest(uint16_t duration);

  protected:

    void sendCommand(uint8_t func, uint8_t strength, uint16_t duration);

    uint16_t mId;
    uint8_t mChannel;
    uint16_t* mTransmitTimings;
    uint16_t mMaxTransmitTimings;
};
